src/analysis/special_functions/trigonometric.lean

```diff
@@ -1478,88 +1478,332 @@ end
+  arcsin x < y ↔ x < sin y :=
+not_le.symm.trans $ (not_congr $ le_arcsin_iff_sin_le' hy).trans not_le
+
+lemma lt_arcsin_iff_sin_lt {x y : ℝ} (hx : x ∈ Icc (-(π / 2)) (π / 2)) (hy : y ∈ Icc (-1 : ℝ) 1) :
+  x < arcsin y ↔ sin x < y :=
+not_le.symm.trans $ (not_congr $ arcsin_le_iff_le_sin hy hx).trans not_le
+
+lemma lt_arcsin_iff_sin_lt' {x y : ℝ} (hx : x ∈ Ico (-(π / 2)) (π / 2)) :
+  x < arcsin y ↔ sin x < y :=
+not_le.symm.trans $ (not_congr $ arcsin_le_iff_le_sin' hx).trans not_le
+
+lemma arcsin_eq_iff_eq_sin {x y : ℝ} (hy : y ∈ Ioo (-(π / 2)) (π / 2)) :
+  arcsin x = y ↔ x = sin y :=
+by simp only [le_antisymm_iff, arcsin_le_iff_le_sin' (mem_Ico_of_Ioo hy),
+  le_arcsin_iff_sin_le' (mem_Ioc_of_Ioo hy)]
+
+@[simp] lemma arcsin_nonneg {x : ℝ} : 0 ≤ arcsin x ↔ 0 ≤ x :=
+(le_arcsin_iff_sin_le' ⟨neg_lt_zero.2 pi_div_two_pos, pi_div_two_pos.le⟩).trans $ by rw [sin_zero]
+
+@[simp] lemma arcsin_nonpos {x : ℝ} : arcsin x ≤ 0 ↔ x ≤ 0 :=
+neg_nonneg.symm.trans $ arcsin_neg x ▸ arcsin_nonneg.trans neg_nonneg
+
+@[simp] lemma arcsin_eq_zero_iff {x : ℝ} : arcsin x = 0 ↔ x = 0 :=
+by simp [le_antisymm_iff]
+
+@[simp] lemma zero_eq_arcsin_iff {x} : 0 = arcsin x ↔ x = 0 :=
+eq_comm.trans arcsin_eq_zero_iff
+
+@[simp] lemma arcsin_pos {x : ℝ} : 0 < arcsin x ↔ 0 < x :=
+lt_iff_lt_of_le_iff_le arcsin_nonpos
+
+@[simp] lemma arcsin_lt_zero {x : ℝ} : arcsin x < 0 ↔ x < 0 :=
+lt_iff_lt_of_le_iff_le arcsin_nonneg
+
+@[simp] lemma arcsin_lt_pi_div_two {x : ℝ} : arcsin x < π / 2 ↔ x < 1 :=
+(arcsin_lt_iff_lt_sin' (right_mem_Ioc.2 $ neg_lt_self pi_div_two_pos)).trans $
+  by rw sin_pi_div_two
+
+@[simp] lemma neg_pi_div_two_lt_arcsin {x : ℝ} : -(π / 2) < arcsin x ↔ -1 < x :=
+(lt_arcsin_iff_sin_lt' $ left_mem_Ico.2 $ neg_lt_self pi_div_two_pos).trans $
+  by rw [sin_neg, sin_pi_div_two]
+
+@[simp] lemma arcsin_eq_pi_div_two {x : ℝ} : arcsin x = π / 2 ↔ 1 ≤ x :=
+⟨λ h, not_lt.1 $ λ h', (arcsin_lt_pi_div_two.2 h').ne h, arcsin_of_one_le⟩
+
+@[simp] lemma pi_div_two_eq_arcsin {x} : π / 2 = arcsin x ↔ 1 ≤ x :=
+eq_comm.trans arcsin_eq_pi_div_two
+
+@[simp] lemma pi_div_two_le_arcsin {x} : π / 2 ≤ arcsin x ↔ 1 ≤ x :=
+(arcsin_le_pi_div_two x).le_iff_eq.trans pi_div_two_eq_arcsin
+
+@[simp] lemma arcsin_eq_neg_pi_div_two {x : ℝ} : arcsin x = -(π / 2) ↔ x ≤ -1 :=
+⟨λ h, not_lt.1 $ λ h', (neg_pi_div_two_lt_arcsin.2 h').ne' h, arcsin_of_le_neg_one⟩
+
+@[simp] lemma neg_pi_div_two_eq_arcsin {x} : -(π / 2) = arcsin x ↔ x ≤ -1 :=
+eq_comm.trans arcsin_eq_neg_pi_div_two
+
+@[simp] lemma arcsin_le_neg_pi_div_two {x} : arcsin x ≤ -(π / 2) ↔ x ≤ -1 :=
+(neg_pi_div_two_le_arcsin x).le_iff_eq.trans arcsin_eq_neg_pi_div_two
+
+lemma maps_to_sin_Ioo : maps_to sin (Ioo (-(π / 2)) (π / 2)) (Ioo (-1) 1) :=
+λ x h, by rwa [mem_Ioo, ← arcsin_lt_pi_div_two, ← neg_pi_div_two_lt_arcsin,
+  arcsin_sin h.1.le h.2.le]
+
+/-- `real.sin` as a `local_homeomorph` between `(-π / 2, π / 2)` and `(-1, 1)`. -/
+@[simp] def sin_local_homeomorph : local_homeomorph ℝ ℝ :=
```

Comment:
It would be nice to generalize this and put it in `topology.local_homeomorph`, as a general construction of a local homeomorph between two types, from a homeomorph (here `sin_order_iso.to_homeomorph`) between subtypes.

Reply:
Note that

* `to_fun` and `inv_fun` must be defined everywhere;
* `sin_order_iso.to_homeomorph` is a `homeomorph` between closed intervals, not open intervals.

Reply:
I don't think the first point matters, because a local homeomorph doesn't care about the values of the function outside the domain; it would be safe to define it as `default` there.  I can think of various ways of addressing the second point, but they're all a bit awkard ... so fair enough, let's leave it.

Reply:
We are going to use this `local_homeomorph` to prove something about derivatives of `arcsin`, so we want `inv_fun` to be definitionally equal to `arcsin`.